src/App.tsx

```diff
@@ -1,26 +1,285 @@
+        });
+      });
+    }
+
+    setIsLoading(true);
+    setShouldFocus(false);
+    setTogglingTodoIds(uncompletedTodos.map(todo => todo.id));
+
+    try {
+      const results = await Promise.allSettled(updatePromises);
+
+      const successfulUpdates = results
+        .map((result, index) => {
+          if (result.status === 'fulfilled' && uncompletedTodos.length === 0) {
+            return todos[index];
+          } else if (
+            result.status === 'fulfilled' &&
+            uncompletedTodos.length !== 0
+          ) {
+            return uncompletedTodos[index];
+          } else {
+            return null;
+          }
+        })
+        .filter(todo => todo !== null);
+
+      setTodos(prevTodos =>
+        prevTodos.map(todo =>
+          successfulUpdates.some(updatedTodo => updatedTodo.id === todo.id)
+            ? { ...todo, completed: !todo.completed }
+            : todo,
+        ),
+      );
+
+      const hasError = results.some(result => result.status === 'rejected');
+
+      if (hasError) {
+        setErrorMessage(ErrorMessages.ErrorToUpdate);
+      }
+    } catch {
+      setErrorMessage(ErrorMessages.ErrorToUpdate);
+    } finally {
+      setIsLoading(false);
+      setShouldFocus(true);
+      setTogglingTodoIds([]);
+    }
+  };
+
+  const handleRenameTodo = async (todo: Todo, title: string) => {
+    setIsLoading(true);
+    setTargetTodoId(todo.id);
+
+    try {
+      const renamedTodo = await updateTodo({
+        ...todo,
+        title,
+      });
+
+      setTodos(currentTodos => {
+        return currentTodos.map(currntTodo =>
+          currntTodo.id === renamedTodo.id ? renamedTodo : currntTodo,
+        );
+      });
+
+      return true;
+    } catch {
+      setErrorMessage(ErrorMessages.ErrorToUpdate);
+
+      return false;
+    } finally {
+      setIsLoading(false);
+      setTargetTodoId(null);
+    }
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <Todos
+        todos={todos}
+        tempTodo={tempTodo}
+        isLoading={isLoading}
+        targetTodoId={targetTodoId}
+        togglingTodoIds={togglingTodoIds}
+        onEmptyTitleError={handleEmptyErrorMessage}
+        onAddTodo={handleAddTodo}
+        onDelete={handleDeleteTodo}
+        onClearCompleted={handleClearCompleted}
+        shouldFocus={shouldFocus}
+        onToggleOneCompleted={handleToggleOneCompleted}
+        onToggleAllCompleted={handleToggleAllCompleted}
+        onRenameTodo={handleRenameTodo}
```

Comment:
There are a lot of props here, try to optimize it and think about whether you need so many props